storage/storage_test.go

```diff
@@ -1417,76 +1420,90 @@ func TestOperationsWithEndpoint(t *testing.T) {
 
-			// Create a bucket
-			c.Bucket("test-bucket").Create(ctx, "pid", nil)
-			if err != nil {
-				t.Errorf("%s: %v", tc.desc, err)
-			}
-			createBucketEndpoint := lastRequestEndpoint
-
-			// Upload an object
-			w := c.Bucket("test-bucket").Object("file").NewWriter(ctx)
-			_, err = io.Copy(w, strings.NewReader("copyng into bucket"))
-			if err != nil {
-				t.Errorf("%s: %v", tc.desc, err)
-			}
-			if closeErr := w.Close(); closeErr != nil {
-				t.Errorf("%s: %v", tc.desc, closeErr)
-			}
-
-			// Download an object
-			rc, err := c.Bucket("test-bucket").Object("file").NewReader(ctx)
-			if err != nil {
-				t.Errorf("%s: %v", tc.desc, err)
-			}
-			defer rc.Close()
-
-			_, err = io.Copy(ioutil.Discard, rc)
-			if err != nil {
-				t.Errorf("%s: %v", tc.desc, err)
-			}
-
-			// Delete bucket
-			if err := c.Bucket("test-bucket").Delete(ctx); err != nil {
-				t.Errorf("%s: %v", tc.desc, err)
-			}
-
-			// Create a bucket
-			c.Bucket("test-bucket").Create(ctx, "pid", nil)
-			if err != nil {
-				t.Errorf("%s: %v", tc.desc, err)
+			operations := []struct {
+				desc       string
+				runOp      func() error
+				wantURL    string
+				wantMethod string
+			}{
+				{
+					desc: "Create a bucket",
+					runOp: func() error {
+						return c.Bucket("test-bucket").Create(ctx, "pid", nil)
+					},
+					wantURL:    "/storage/v1/b?alt=json&prettyPrint=false&project=pid",
+					wantMethod: "POST",
+				},
+				{
+					desc: "Upload an object",
+					runOp: func() error {
+						w := c.Bucket("test-bucket").Object("file").NewWriter(ctx)
+						_, err = io.Copy(w, strings.NewReader("copyng into bucket"))
+						if err != nil {
+							return err
+						}
+						return w.Close()
+					},
+					wantURL:    "/upload/storage/v1/b/test-bucket/o?alt=json&name=file&prettyPrint=false&projection=full&uploadType=multipart",
+					wantMethod: "POST",
+				},
+				{
+					desc: "Download an object",
+					runOp: func() error {
+						rc, err := c.Bucket("test-bucket").Object("file").NewReader(ctx)
+						if err != nil {
+							return err
+						}
+
+						_, err = io.Copy(ioutil.Discard, rc)
+						if err != nil {
+							return err
+						}
+						return rc.Close()
+					},
+					wantURL:    "/test-bucket/file",
+					wantMethod: "GET",
+				},
+				{
+					desc: "Delete bucket",
+					runOp: func() error {
+						return c.Bucket("test-bucket").Delete(ctx)
+					},
+					wantURL:    "/storage/v1/b/test-bucket?alt=json&prettyPrint=false",
+					wantMethod: "DELETE",
+				},
 			}
 
-			if lastRequestEndpoint != createBucketEndpoint {
-				t.Errorf("unexpected change in endpoint for creating a bucket\n\tgot:\t\t%v\n\toriginal:\t%v", lastRequestEndpoint, createBucketEndpoint)
+			// Check that the calls made to the server are as expected
+			// given the operations performed
+			for _, op := range operations {
+				if err := op.runOp(); err != nil {
+					t.Errorf("%s: %v", op.desc, err)
+				}
+				u, method := <-gotURL, <-gotMethod
```

Comment:
What happens if the channel never sends? Should we add some kind of timeout to each subtest?

Reply:
Good point, I added a timeout.